### PR TITLE
8344177: Remove SecurityManager and related calls from java.management

### PR DESCRIPTION
**Pull request by kevinjwalls**
Remove redundant SecurityManager, AccessController references
(following on from JDK-8338411: Implement JEP 486: Permanently Disable the Security Manager).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344177](https://bugs.openjdk.org/browse/JDK-8344177): Remove SecurityManager and related calls from java.management (**Enhancement** - P4)

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22100/head:pull/22100` \
`$ git checkout pull/22100`

Update a local copy of the PR: \
`$ git checkout pull/22100` \
`$ git pull https://git.openjdk.org/jdk.git pull/22100/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22100`

View PR using the GUI difftool: \
`$ git pr show -t 22100`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22100.diff">https://git.openjdk.org/jdk/pull/22100.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22100#issuecomment-2482729892)
</details>

### CODE REVIEW

File: src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java [hunk 1]
```diff
@@ -74,13 +72,9 @@
...
- * <p>If there is no SecurityManager, then the access controller will refuse
- * to create an MBean that is a ClassLoader.  This prevents
- * people from opening security holes unintentionally. Otherwise, it
- * would not be obvious that granting write access grants the ability to
- * download and execute arbitrary code in the target MBean server. Advanced
- * users who do want an MBean which is a ClassLoader are presumably advanced enough
- * to handle policy files and security managers.</p>
+ * <p>An MBean could possibly be a CassLoader.
+ * Users who do want an MBean which could be a ClassLoader should not permit
+ * MBean creation.</p>
```
Comment (dfuch) [inline src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java:77]: Did you mean "who do not want" ?

Comment (kevinjwalls) [reply src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java:77]: A CassLoader also, will clarify! 8-)

File: src/java.management/share/classes/javax/management/remote/JMXConnectorFactory.java [hunk 2]
```diff
@@ -447,20 +440,12 @@ static <T> T getProvider(JMXServiceURL serviceURL,
...
-                        return super.loadClass(name, resolve);
-                    }
-                };
-            }
-        }) : null;
+        return parent != null ? new ClassLoader(parent) {
+                                    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException
+                                        { return super.loadClass(name, resolve); }
+                                }
+                              : null;
```
Comment (dfuch) [inline src/java.management/share/classes/javax/management/remote/JMXConnectorFactory.java:448]: I suspect we could just return `parent` here but that might require some deeper investigation.

File: src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java [hunk 3]
```diff
@@ -174,15 +168,8 @@ public ObjectInstance createMBean(String className, ObjectName name)
         MBeanException,
         NotCompliantMBeanException {
         checkCreate(className);
-        @SuppressWarnings("removal")
-        SecurityManager sm = System.getSecurityManager();
-        if (sm == null) {
-            Object object = getMBeanServer().instantiate(className);
-            checkClassLoader(object);
```
Comment (dfuch) [inline src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java:181]: I believe we should keep this call to `checkClassLoader(object)` in place for now. What to do in this case will require some more thinking - I'd suggest handling that part in a separate PR.

File: src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java [hunk 4]
```diff
@@ -74,13 +72,8 @@
...
  *
- * <p>If there is no SecurityManager, then the access controller will refuse
- * to create an MBean that is a ClassLoader.  This prevents
- * people from opening security holes unintentionally. Otherwise, it
- * would not be obvious that granting write access grants the ability to
- * download and execute arbitrary code in the target MBean server. Advanced
- * users who do want an MBean which is a ClassLoader are presumably advanced enough
- * to handle policy files and security managers.</p>
+ * <p>An MBean could possibly be a ClassLoader.  Users who do not want an
+ * MBean which could be a ClassLoader should not permit MBean creation.</p>
```
Comment (dfuch) [inline src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java:76]: Maybe we should just refuse to create an MBean which is a ClassLoader for now (see further comments below) and handle that part in a dedicated PR.

Comment (kevinjwalls) [reply src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java:76]: Yes agreed I'm going to update again with checkClassLoader put back.

File: src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java (hunk 3 above)
Comment (kevinjwalls) [reply src/java.management/share/classes/com/sun/jmx/remote/security/MBeanServerAccessController.java:181]: Yes agreed.

File: src/java.management/share/classes/sun/management/ManagementFactoryHelper.java [hunk 5]
```diff
@@ -191,18 +187,13 @@ static boolean isAvailable() {
...
-                    if (logging.isPresent()) {
-                        return Class.forName(logging.get(), className);
-                    }
-                    return null;
-                }
-            });
+            Optional<Module> logging = ModuleLayer.boot().findModule("java.logging");
+            if (logging.isPresent()) {
+                return Class.forName(logging.get(), className);
+            }  else {
```
Comment (turbanoff) [inline src/java.management/share/classes/sun/management/ManagementFactoryHelper.java:194]:
```suggestion
            } else {
```

File: src/java.management/share/classes/sun/management/VMManagementImpl.java [hunk 6]
```diff
@@ -255,8 +246,7 @@ private synchronized PerfInstrumentation getPerfInstrumentation() {
         }
 
         // construct PerfInstrumentation object
-        @SuppressWarnings("removal")
-        Perf perf =  AccessController.doPrivileged(new Perf.GetPerfAction());
+        Perf perf =  Perf.getPerf();
```
Comment (turbanoff) [inline src/java.management/share/classes/sun/management/VMManagementImpl.java:249]:
```suggestion
        Perf perf = Perf.getPerf();
```

File: src/java.management/share/classes/sun/management/ManagementFactoryHelper.java (hunk 5 above)
Comment (kevinjwalls) [reply src/java.management/share/classes/sun/management/ManagementFactoryHelper.java]: ok thanks

File: src/java.management/share/classes/sun/management/VMManagementImpl.java (hunk 6 above)
Comment (kevinjwalls) [reply src/java.management/share/classes/sun/management/VMManagementImpl.java]: ok yes, an existing double-space but we'll fix it 8-)

**Comment by AlanBateman** [comment]
Maybe follow-up, but can you check the usage of ReflectUtil.forName in MBeanServerFactory.loadBuilderClass. It looks like it wants to try the boot loader when the CCL is not set, in which case this can be changed to use the 3-arg forName a loader of null. If we can make this go away then I think it may be the last usage of ReflectUtil by the java.management module. It would be nice if we could make the qualified export of  sun.reflect.misc to java.management go away if we can.

**Comment by kevinjwalls** [comment]
> Maybe follow-up, but can you check the usage of ReflectUtil.forName in MBeanServerFactory.loadBuilderClass. 

Happy to do that, yes as it may change module xports maybe it's better I look at that separately.

Actually I've missed a few ReflectUtil imports here in java.management that can go, will update.

**Comment by AlanBateman** [comment]
Does the latest update mean that java.base's module-info can be updated to not export sun.reflect.misc to java.management?

**Comment by kevinjwalls** [comment]
> Does the latest update mean that java.base's module-info can be updated to not export sun.reflect.misc to java.management?

Not quite, there is use of  sun.reflect.misc.MethodUtil and sun.reflect.misc.ConstructorUtil by java.management.  A few places call 
 MethodUtil.invoke().j


java.management.rmi will be done with its use of sun.reflect.misc after the separate change 8344394.

**Comment by kevinjwalls** [comment]
A few updates.

Triggered by investigating this, I've logged these future cleanup items (not for jdk-24):

8344966: Remove the allowNonPublic MBean compatibility property
8344969: Remove the jmx.mxbean.multiname compatibility property
8344976: Remove the jmx.invoke.getters compatibility property
8345045: Remove the jmx.remote.x.buffer.size JMX notification property
8345048: Remove the jmx.extend.open.types compatibility property
8345049: Remove the jmx.tabular.data.hash.map compatibility property

**Comment by kevinjwalls** [comment]
Merged with master after PR 22478 is integrated - that had the same changes for src/java.management/share/classes/sun/management/VMManagementImpl.java
(that file is still appears in this PR with a change, as I'd removed another of the existing extra spaces).

File: src/java.management/share/classes/javax/management/MBeanServerFactory.java [hunk 7]
```diff
@@ -420,7 +365,7 @@ private static Class<?> loadBuilderClass(String builderClassName)
         }
 
         // No context class loader? Try with Class.forName()
-        return ReflectUtil.forName(builderClassName);
+        return Class.forName(builderClassName);
```
Comment (dfuch) [inline src/java.management/share/classes/javax/management/MBeanServerFactory.java:368]: Maybe we should keep that call to ReflectUtil here for the moment?

Comment (AlanBateman) [reply src/java.management/share/classes/javax/management/MBeanServerFactory.java:368]: The java.management module is mapped to the boot loader so it's okay to replace the use of ReflectUtil.forName with Class.forName here (no behavior change).

**Review by dfuch — COMMENTED** [review]
Globally LGTM - except I believe we should keep that one call to ReflectUtil::forName for now.

File: src/java.management/share/classes/javax/management/MBeanServerFactory.java (hunk 7 above)
Comment (dfuch) [reply src/java.management/share/classes/javax/management/MBeanServerFactory.java:368]: OK - I withdraw my objection then

**Review by dfuch — APPROVED** [review]

**Comment by kevinjwalls** [comment]
Thanks Daniel, and thanks for the discussion above.

**Pull request closed without merging**